Add a command snippet to build only modified documentation

Maarrk opened the pull request:
I imagine the main reason to use `FILELIST` is to only rebuild the files that are being worked on.

I don't usually work in ~~PowerShell~~ Windows but if you require a ~~PS~~ Windows version to accept the contribution, I can try to make it work.

There are some more nice ways to use it like `git diff --staged` or `git diff master` but I thought it would be too much here. But let me know if you'd actually like it included, I imagine it could help reviewers.

<!--
Please target the `master` branch in priority.
PRs can target other branches (e.g. `3.2`, `3.5`) if the same change was done in `master`, or is not relevant there.
PRs must not target `stable`, as that branch is updated manually.

The type of content accepted into the documentation is explained here:
https://docs.godotengine.org/en/latest/community/contributing/content_guidelines.html
-->

skyace65 commented on contributing/documentation/building_the_manual.rst:
```diff
@@ -167,3 +167,10 @@ You can specify a list of files to build, which can greatly speed up compilation
+
+.. code:: sh
+
+    FILELIST="\"$(git diff HEAD --name-only | sed -z 's/\n/\" \"/')\"" make html
```
the ``make html`` part needs to be at the front, I'm also getting this error on windows in command prompt ``'sed' is not recognized as an internal or external command,`` It also gives me an error with ``sed`` in PowerShell

Also, not a issue specific to this PR but I'm not even sure if the ``FILELIST`` command actually works, I tried the above example and it just builds the entire docs

tetrapod00 replied:
`FILELIST` definitely works in general, I tested just now on Ubuntu via WSL.

Maarrk replied:
I'm sorry but on Windows I'm unable to successfully specify the FILELIST in any way. Even not doing the git trickery, it outputs classes whatever I try in PowerShell and Command Prompt, using an environment variable, before html, after html etc.

It's already a wonder that something resembling `make` works at all in Windows. If more people reproduce @skyace65's results, maybe we should mark the whole `FILELIST` section as non-Windows? (I don't count myself because I know next to nothing about Windows commandline usage)

tetrapod00 replied:
AFAIK `set SPHINXOPTS=-j2 && make html` doesn't work on Windows either, but I can't track down the issue that confirms that. (It's why I switched to WSL for docs builds.)

We may want to suggest using WSL as a general purpose solution for users who frequently build the docs on windows.

skyace65 replied:
From what I'm seeing ``sed`` is only on Linux. I'm fine with adding this to the docs but we need to make sure it's clear that command will only work on Linux.

Maarrk replied:
I know about `sed` only being a Linux thing. There is a counterpart on Windows, that's not the issue.

The question now is whether you're able to get the FILELIST to work on Windows even passing the filenames manually

tetrapod00 replied:
I can confirm the inability to set FILELIST on windows; tested in both cmd and powershell and before and after the `make html` or `.\make html` (in powershell). I trust Maarrk's testing here. I also tried `set FILELIST='about.rst'` since that is recommended for setting SPHINXOPTS and that doesn't work either.

If only the basic `make html` command works on windows, I think it's good to add a warning about that and also suggest WSL if you're building often.

Maarrk commented:
> Also, not a issue specific to this PR but I'm not even sure if the FILELIST command actually works, I tried the example above what you added in this PR and it just builds the entire docs

> I'm fine with adding this to the docs but we need to make sure it's clear that command will only work on Linux.

> We may want to suggest using WSL as a general purpose solution for users who frequently build the docs on windows.

I had a look at the `make.bat` script, and it doesn't handle `FILELIST` in any way. So I clarified that with a `.. warning` with WSL suggestion.

https://github.com/godotengine/godot-docs/blob/9abcbd707e35d4137cb70321b1e795794e7c9d0f/make.bat

I also rewrote the explanation and modified the snippet to hopefully make it a bit easier to follow

tetrapod00 commented:
Oh, and to be clear I tested the commands on Ubuntu via WSL and they work well. Personally the `master` version is going to be the most useful for me, to test PRs, so I recommend adding that as a copyable snippet. 

Worth noting that it *can* pull in invalid files like images, but that gets filtered out (and writes warnings to the console). Also if the PR hasn't been recently rebased against master it can pull in some irrelevant files. But overall it should still be a lot faster than building the whole docs, and easier than adding in the relevant pages manually in some cases

tetrapod00 commented on contributing/documentation/building_the_manual.rst:
```diff
@@ -162,8 +162,29 @@ RAM for Sphinx alone.
+You can replace ``HEAD`` with ``master`` to return all files changed from the
+``master`` branch.
+
+.. code:: sh
+    make html FILELIST="$(git diff master --name-only | sed -z 's/\n/ /g')"
```
```suggestion
.. code:: sh

    make html FILELIST="$(git diff master --name-only | sed -z 's/\n/ /g')"
```
My bad, I gave you a bad RST suggestion

Maarrk replied:
fixed

Maarrk commented:
> Worth noting that it can pull in invalid files like images, but that gets filtered out (and writes warnings to the console). 

@tetrapod00 can you have a look while you're here?

tetrapod00 reviewed (approved):
I think this looks good now! Seems like a useful addition.

tetrapod00 commented on contributing/documentation/building_the_manual.rst:
```diff
@@ -162,8 +162,26 @@ RAM for Sphinx alone.
     make html FILELIST='classes/class_node.rst classes/class_resource.rst'
+
+The list of files can also be read from ``FILELIST`` environment variable.
+This way you can automatically get names of all files that were changed since
```
```suggestion
This way you can automatically get the names of all files that have changed since
```

skyace65 commented:
Thanks!
